File: Vector.py
class Vector:
    x, y, z = 0, 0, 0;
    formatString = "(%.2f, %.2f, %.2f)"

    # constructors
    def __init__(self, x = 0, y = 0, z = 0):
        self.init(x, y, z)
        
    def init(self, x, y, z):
        self.x, self.y, self.z = x, y, z

    # magnitudes
    def length(self):
        return (self.x ** 2 + self.y ** 2 + self.z ** 2) ** 0.5

    def lengthSquared(self):
        return self.x ** 2 + self.y ** 2 + self.z ** 2

    def length2d(self):
        return (self.x ** 2 + self.y ** 2) ** 0.5

    def lengthSquared2d(self):
        return self.x ** 2 + self.y ** 2;

    # operators
    def dot(self, other):
        assert isinstance(other, Vector), "may only dot with another Vector"
        return self.x * other.x + self.y * other.y + self.z * other.z
        
    def add(self, *args):
        if len(args) == 1 and isinstance(args[0], Vector):
            other = args[0]
            return Vector(self.x + other.x, self.y + other.y, self.z + other.z)
        elif len(args) == 3 and isinstance(args[0], (int, float)) and isinstance(args[1], (int, float)) and isinstance(args[2], (int, float)):
            return Vector(self.x + args[0], self.y + args[1], self.z + args[2])
        else:
            raise TypeError("unknown add arguments, expected [Vector other] or [number x, number y, number z]")
            
    def iadd(self, *args):
        if len(args) == 1 and isinstance(args[0], Vector):
            other = args[0]
            self.x += other.x
            self.y += other.y
            self.z += other.z
        elif len(args) == 3 and isinstance(args[0], (int, float)) and isinstance(args[1], (int, float)) and isinstance(args[2], (int, float)):
            self.x += args[0]
            self.y += args[1]
            self.z += args[2]
        else:
            raise TypeError("unknown iadd arguments, expected [Vector other] or [number x, number y, number z]")
    
    def sub(self, *args):
        if len(args) == 1 and isinstance(args[0], Vector):
            other = args[0]
            return Vector(self.x - other.x, self.y - other.y, self.z - other.z)
        elif len(args) == 3 and isinstance(args[0], (int, float)) and isinstance(args[1], (int, float)) and isinstance(args[2], (int, float)):
            return Vector(self.x - args[0], self.y - args[1], self.z - args[2])
        else:
            raise TypeError("unknown sub arguments, expected [Vector other] or [number x, number y, number z]")
            
    def isub(self, *args):
        if len(args) == 1 and isinstance(args[0], Vector):
            other = args[0]
            self.x -= other.x
            self.y -= other.y
            self.z -= other.z
        elif len(args) == 3 and isinstance(args[0], (int, float)) and isinstance(args[1], (int, float)) and isinstance(args[2], (int, float)):
            self.x -= args[0]
            self.y -= args[1]
            self.z -= args[2]
        else:
            raise TypeError("unknown isub arguments, expected [Vector other] or [number x, number y, number z]")
    
    def scale(self, f):
        assert isinstance(f, (int, float)), "must scale by a number"
        return Vector(self.x * f, self.y * f, self.z * f)
        
    def iscale(self, f):
        assert isinstance(f, (int, float)), "must iscale by a number"
        self.x *= f
        self.y *= f
        self.z *= f

    # util
    def setOutputPercision(self, decimalPlaces):
        assert isinstance(decimalPlaces, int), "setOutputPercision takes an integer, the number of decimals to display when converting to string. -1 = all decimals"
        if decimalPlaces < 0 or decimalPlaces > 16:
            decimalPlaces = 16
        self.formatString = "(%%.%df, %%.%df, %%.%df)" % (decimalPlaces, decimalPlaces, decimalPlaces)
    
    def __str__(self):
        return self.formatString % (self.x, self.y, self.z)
    
    def unit(self):
        length = self.length()
        return Vector(0.5773502691896258, 0.5773502691896258, 0.5773502691896258) if length == 0 else (Vector(self) if length == 1 else self.scale(1/length))
    

File: Vector.java
public class Vector
{
    private String formatString = "(%.2f, %.2f, %.2f)";
    public double x, y, z = 0;

    // constructors
    Vector(){}
    Vector(double x, double y, double z)
    {
        this.x = x;
        this.y = y;
        this.z = z;
    }
    Vector(int x, int y, int z)
    {
        this.x = (double)x;
        this.y = (double)y;
        this.z = (double)z;
    }
    Vector(Vector other)
    {
        this.x = other.x;
        this.y = other.y;
        this.z = other.z;
    }
    void init(Vector other)
    {
        this.x = other.x;
        this.y = other.y;
        this.z = other.z;
    }
    void init(double x, double y, double z)
    {
        this.x = x;
        this.y = y;
        this.z = z;
    }
    void init(int x, int y, int z)
    {
        this.x = (double)x;
        this.y = (double)y;
        this.z = (double)z;
    }

    // magnitudes
    double length()
    {
        return Math.sqrt(x*x + y*y + z*z);
    }

    double lengthSquared()
    {
        return x*x + y*y + z*z;
    }

    double length2d()
    {
        return Math.sqrt(x*x + y*y);
    }

    double lengthSquared2d()
    {
        return x*x + y*y;
    }

    // operators
    double dot(Vector other)
    {
        return x * other.x + y * other.y + z * other.z;
    }
    Vector add(Vector other)
    {
        return new Vector(x + other.x, y + other.y, z + other.z);
    }
    Vector add(double x, double y, double z)
    {
        return new Vector(this.x + x, this.y + y, this.z + z);
    }
    Vector add(int x, int y, int z)
    {
        return new Vector(this.x + (double)x, this.y + (double)y, this.z + (double)z);
    }
    void iadd(Vector other)
    {
        x += other.x;
        y += other.y;
        z += other.z;
    }
    void iadd(double x, double y, double z)
    {
        this.x += x;
        this.y += y;
        this.z += z;
    }
    void iadd(int x, int y, int z)
    {
        this.x += (double)x;
        this.y += (double)y;
        this.z += (double)z;
    }
    Vector sub(Vector other)
    {
        return new Vector(x - other.x, y - other.y, z - other.z);
    }
    Vector sub(double x, double y, double z)
    {
        return new Vector(this.x - x, this.y - y, this.z - z);
    }
    Vector sub(int x, int y, int z)
    {
        return new Vector(this.x - (double)x, this.y - (double)y, this.z - (double)z);
    }
    void isub(Vector other)
    {
        x -= other.x;
        y -= other.y;
        z -= other.z;
    }
    void isub(double x, double y, double z)
    {
        this.x -= x;
        this.y -= y;
        this.z -= z;
    }
    void isub(int x, int y, int z)
    {
        this.x -= (double)x;
        this.y -= (double)y;
        this.z -= (double)z;
    }
    Vector scale(double f)
    {
        return new Vector(x * f, y * f, z * f);
    }
    Vector scale(int f)
    {
        return new Vector(x * (double)f, y * (double)f, z * (double)f);
    }
    void iscale(double f)
    {
        x *= f;
        y *= f;
        z *= f;
    }
    void iscale(int f)
    {
        x *= (double)f;
        y *= (double)f;
        z *= (double)f;
    }

    // util
    void setOutputPercision(int decimalPlaces)
    {
        if (decimalPlaces < 0 || decimalPlaces > 16)
            decimalPlaces = 16;
        this.formatString = String.format("(%%.%df, %%.%df, %%.%df)", decimalPlaces, decimalPlaces, decimalPlaces);
    }
    @Override
    public String toString()
    {
        return String.format(this.formatString, x, y, z);
    }
    Vector unit()
    {
        double length = this.length();
        // tfw you can't switch on a double
        return length == 0.f ? new Vector(0.5773502691896258, 0.5773502691896258, 0.5773502691896258) : length == 1.f ? new Vector(this) : this.scale(1 / length);
    }
}
